### ISSUE TITLE
Regreession causing false positive in CA1416

### ISSUE THREAD

**Issue by jhudsoncedaron** [dotnet/roslyn-analyzers#5363]
### Analyzer

**Diagnostic ID**: [CA1416](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1416): Xyz is only available on Windows

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.302](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

There's a regression in Roslyn analyzers where it is no longer willing to check the callsites for a `private static` method. The platform verification should quite often be one step up the stack to avoid attempting to run static constructors for classes that are only supported on certain platforms.

This code did not alert when .NET 5 was first released but is alerting now.

### Steps To Reproduce

````
public namsespace Something {
    public class SomeClass {
        public void SomeFunction()
        {
             RequireSomeInfo(); /* lazy loaded */
             /* Does something with newly loaded static info */
        }

        private static void RequireSomeInfo()
        {
            if (SomeInfoLoaded) return;
            if (System.Runtime.InteropServices.RuntimeInformation.IsOSPlatform(System.Runtime.InteropServices.OSPlatform.Windows))
                RequireSomeInfoRegistry();
            else
                RequireSomeInfoFile();
            SomeInfoLoaded = true;
        }

        private static void RequireSomeInfoRegistry()
        {
            /* This line raises CA1416 */
            using var w = Microsoft.Win32.Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\Windows\CurrentVersion");
            /* Reads omitted for brevity */
        }

        private static void RequireSomeInfoFile()
        { /* Omitted for brevity */ }
    }
}
````

### Expected behavior

NO CA1416 raised because the code is only reachable on Windows

### Actual behavior

CA1416 is raised annoying the developer

**Comment by buyaa-n** [comment]
> ### Expected behavior
>
> NO CA1416 raised because the code is only reachable on Windows

@jhudsoncedaron not sure why you were not getting a warning in 5.0 but the warning is expected because `RequireSomeInfoRegistry` is using windows only API you should annotate the method with `SupportedOSPlatform("windows")` attribute then there will be no warning

```cs
SupportedOSPlatform("windows")
private static void RequireSomeInfoRegistry()
{
     /* should be no warning after adding attribute */
    using var w = Microsoft.Win32.Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\Windows\CurrentVersion");
}
```

**Comment by jhudsoncedaron** [comment]
What a poor reflection on the analyzer.

**Comment by Evangelink** [comment]
@buyaa-n Is there any work you want us to do here?

**Comment by buyaa-n** [comment]
> What a poor reflection on the analyzer.

`RequireSomeInfoRegistry()` should be annotated `SupportedOSPlatform("windows")` so that callers of this API was aware of it is windows only, it will be very expensive for the analyzer to try to check all callers of the method, annotating the method is forwarding the requirements to the callers and it is the most efficient way of handling this

> @buyaa-n Is there any work you want us to do here

No, thanks, it is by design
